### PR TITLE
Median and average artefact

### PR DESCRIPTION
**Pull request by yger**
Since this was something in SpyKING CIRCUS, I would like to have again the possibility, in order to remove artefacts, to subtract either the mean or the median computed over several artefacts. This is what this PR is about. It offers two new possibilities for artefact removal: mean or median, as explained in the docstrings.

### CODE REVIEW

**Comment by alejoe91** [comment]
Any chance we can extend it so that: 
1. the artifact template is computed
2. for each artifact the avg template is scaled

what do you think?

**Comment by yger** [comment]
Not sure to understand. 
1. The average template is computed in my PR, via a waveformExtractor object. We need it to subtract it. 
2. Not getting also the second point: You mean that you want, for every artefact, to find an amplitude factor for the subtraction in order to scale it? Why would you do that?

**Comment by yger** [comment]
Not sure to understand. 
1. The average template is computed in my PR, via a waveformExtractor object. We need it to subtract it. 
2. Not getting also the second point: You mean that you want, for every artefact, to find an amplitude factor for the subtraction in order to scale it? Why would you do that?

**Comment by yger** [comment]
Not sure to understand. 
1. The average template is computed in my PR, via a waveformExtractor object. We need it to subtract it. 
2. Not getting also the second point: You mean that you want, for every artefact, to find an amplitude factor for the subtraction in order to scale it? Why would you do that?

**Comment by yger** [comment]
The only thing is that in SC, we could pass a list of triggers and a list of stim_types, to be able to compute several artefact templates (one per stim type). This is because quite often, experimentalists have more than one artefact type. I guess the option with SI would be to have a chain of RemoveArtefacts objects, one for each type of artefact

**Comment by alejoe91** [comment]
> Not sure to understand.
> 
> 1. The average template is computed in my PR, via a waveformExtractor object. We need it to subtract it.
> 2. Not getting also the second point: You mean that you want, for every artefact, to find an amplitude factor for the subtraction in order to scale it? Why would you do that?

Because I think there will be some slight variation between artifacts (not massive). At least that's what I observed with electrical stim

**Comment by alejoe91** [comment]
But I guess this could be done later (if needed!)

File: spikeinterface/preprocessing/remove_artifacts.py [hunk 1]
```diff
@@ -89,25 +95,37 @@ def __init__(self, recording, list_triggers, ms_before=0.5, ms_after=3.0, mode='
...
 
+        if mode in ['median', 'average']:
+            sorting = NumpySorting.from_times_labels(list_triggers, [np.ones(len(i)) for i in list_triggers], recording.get_sampling_frequency())
+            waveforms_params = {'ms_before' : ms_before, 'ms_after' : ms_after}
+            w = extract_waveforms(recording, sorting, None, mode='memory', **waveforms_params, return_scaled=False)
+            artefact = w.get_all_templates(mode=mode)[0]
+
+            for seg_index, parent_segment in enumerate(recording._recording_segments):
+                self._recording_segments[seg_index].artefact = artefact
+
```
Comment (alejoe91) [inline spikeinterface/preprocessing/remove_artifacts.py:114]: Can you move this before instantiating the segment, so you can directly pass the `artifact`?

Comment (yger) [reply spikeinterface/preprocessing/remove_artifacts.py]: Indeed, this has been solved

**Review by alejoe91 — APPROVED** [review]
Looks good to me! @samuelgarcia want to take a look?

**Comment by samuelgarcia** [comment]
yep. I will have a short look

**Review by alejoe91 — CHANGES_REQUESTED** [review]
add labels/list of different trigger types

File: spikeinterface/preprocessing/tests/test_remove_artifacts.py [hunk 2]
```diff
@@ -28,6 +29,16 @@ def test_remove_artifacts():
     assert not np.any(traces_short_0)
     assert not np.any(traces_short_1)
 
+    rec_rmart_mean = remove_artifacts(rec, triggers, ms_before=10, ms_after=10, mode='average')
```
Comment (alejoe91) [inline spikeinterface/preprocessing/tests/test_remove_artifacts.py:32]: @yger can you add a test with labels?

**Comment by yger** [comment]
There is one, but mostly to check that this is working, not really to test how good everything is

**Review by alejoe91 — APPROVED** [review]

**Comment by alejoe91** [comment]
merging! Thanks!

**Pull request merged**